crates/randicheck/src/main.rs

```diff
@@ -0,0 +1,327 @@
+// A prototype haskell parser that can parse data types and functions with case statements
+use std::collections::HashMap;
+use streaming_iterator::StreamingIterator;
+use tree_sitter::{Parser, Query, QueryCursor, Tree};
+
+// a haskell data type has a name and a list of constructors
+#[derive(Debug)]
+struct DataType {
+    Name: String,
+    Constructors: Box<[Constructor]>,
+}
+
+//a constructor for an adt is just a list of types
+#[derive(Clone,Debug)]
+struct Constructor {
+    Types: Box<[Type]>,
+}
+// we only support bools and custom types for now (may have to add boolean with a T/F or placeholder later)
+#[derive(Clone,Debug)]
+enum Type {
+    Bool,
+    Custom(String),
+}
+
+// a function has a name, it's input type(s), and it's output type
+#[derive(Debug)]
+struct Function {
+    Name: String,
+    Inputs: Box<[Type]>,
+    Output: Box<Type>,
+    Body: Box<[FunctionBody]>,
+}
+//the body of a function is either a case statement which has what it is matching on and a pattern, or a boolean
+#[derive(Clone,Debug)]
+enum FunctionBody {
+    Case(String,Pattern),
+    Bool,
+}
+// a pattern has a list of types that it matches on and gives a boolean
+#[derive(Clone,Debug)]
+struct Pattern {
+    True_On: Box<[Type]>,
+    gives : Bool}
+
+#[derive(Clone,Debug)]
+enum Bool {
+    True,
+    False,
+}
+
+// the full haskell file is a list of data types and functions, where we only support custom types, bools, and functions with case statements
+#[derive(Debug)]
+struct Haskell {
```

Comment:
I would call this something like Program. our aspiration is that this is not haskell specific.

Reply:
I think that's valid in future, but considering that currently this is only targetting haskell, and when adding additional languages this will likely be refactored to a seperate ast file, pushing a change for one word isn't vitally important. I'll make sure to change it for my next commit though.